### DISCUSSION TITLE
Getting VKT & number of bus operating for each route

### DISCUSSION THREAD

**Discussion by nadiaqamilla** [ipeaGIT/gtfs2emis#104] [Q&A] [answered]
Hi, I am wondering if i could get the vehicle kilometers travelled (VKT) as well as the number of bus operating for each route from GTFS data? I can't seem to find a code reference so far. I read the paper titled Estimating public transport emissions from General Transit Feed Specification data and saw that it's possible to obtain VKT from GTFS.

<img width="583" alt="vkt" src="https://github.com/ipeaGIT/gtfs2emis/assets/140500332/2dda730e-4cb9-4c76-b841-6e463530dff2">

Cheers & thanks before
Nadia

**Comment by Joaobazzo** [comment, marked as answer]
Hi @nadiaqamilla 
The codes we used to reproduce the results from the article can be found in this repository <<https://github.com/ipeaGIT/gtfs2emis_paper>>. 
Two quick comments:
1) Regarding the number of buses in each operating line, in the article we assigned a distribution of buses with the same characteristics as the fleet data of São Paulo. The ideal scenario would be to have a single bus type for each trip_id. However, given the lack of detailed data (e.g. bus ID X assigned to route Y), we assumed that all of São Paulo's buses have equal chances of accessing a certain route. You can access the fleet data in this script <<https://github.com/ipeaGIT/gtfs2emis_paper/blob/main/R/1.0_data-preparation.R>>.

2) You can check the total VKT we estimate running those scripts from <<https://github.com/ipeaGIT/gtfs2emis_paper>>. However, if you are only interested in the total VKT, you can use `gtfstools` R package. See this reproducible example using `gtfs2emis` and `gtfstools` alternatives.

```R

library(gtfs2emis)
library(gtfstools)

# read GTFS
gtfs_file <- system.file("extdata/bra_cur_gtfs.zip", package = "gtfs2emis")
gtfs <- gtfstools::read_gtfs(gtfs_file) 

# keep a single trip_id to speed up this example
gtfs_small <- gtfstools::filter_by_trip_id(gtfs, trip_id ="4451136")

# run transport model
tp_model <- gtfs2emis::transport_model(gtfs_data = gtfs_small,
                            min_speed = 2,
                            max_speed = 80,
                            new_speed = 20,
                            spatial_resolution = 100,
                            parallel = FALSE)

# 1) total VKT from 'transport_model' function
sum(tp_model$dist)

# 2) total VKT using gtfstools
gtfstools::get_trip_length(gtfs_small,trip_id = "4451136")


```

**Reply by nadiaqamilla** [reply]
Thanks a lot @Joaobazzo.
I tried using both scripts using transport_model from gtfs2emis and get_trip_length from gtfstools. 

I am using [Transjakarta BRT GTFS](https://ppid.transjakarta.co.id/pusat-data/data-terbuka/transjakarta-gtfs-feed). I used the code given above to find VKT:

```
# read GTFS
gtfs_file <- system.file("extdata/jkt_gtfs.zip", package = "gtfs2emis")
gtfs <- gtfstools::read_gtfs(gtfs_file) 

# keep a single trip_id to speed up this example
gtfs_small <- gtfstools::filter_by_trip_id(gtfs, trip_id ="7A-R01")

# run transport model
tp_model <- gtfs2emis::transport_model(gtfs_data = gtfs_small,
                                       min_speed = 2,
                                       max_speed = 80,
                                       new_speed = 20,
                                       spatial_resolution = 100,
                                       parallel = FALSE)

# 1) total VKT from 'transport_model' function
sum(tp_model$dist)

# 2) total VKT using gtfstools
gtfstools::get_trip_length(gtfs_small,trip_id = "7A-R01")
```

However, the result showed stark differences between the two approaches. 
<img width="366" alt="image" src="https://github.com/ipeaGIT/gtfs2emis/assets/140500332/2258f637-b3e3-4a0a-bd03-500869b46fbb">

Is it perhaps that the result from transport_model is the daily kilometers travelled?

**Reply by Joaobazzo** [reply]
hi @nadiaqamilla . This is probably because you have multiple trip_id's in your `tp_model`. 
You can check individual VKT per trip_id 

```R
unique(tp_model$trip_id) # verify vector of trip_id's

data.table::setDT(tp_model)[,sum(dist),by=.(trip_id)] # estimate sum of VKT per trip_id
```

**Reply by nadiaqamilla** [reply]
Hi @Joaobazzo 
Apparently it is correct that the GTFS data contains multiple branched id within a trip_id. What does those multiple id(s) mean in the data? Let's say I have one trip_id -- 7A-R01 but there are id for 7A-R01_1 until 7A-R01_73, all yielding the same VKT.
<img width="168" alt="image" src="https://github.com/ipeaGIT/gtfs2emis/assets/140500332/4deaa672-d6d5-4b5f-ad79-b86898e0801b">

Another question, would I be able to run the emission model for each detailed trip_id (7A-R01_1), to get the pollutants data specifically for each trip? I previously run the model by only setting the trip_id of "7A-R01", but thought that the result might not be accurate as it seems to sum the multiple ids' pollutant mass. Please kindly correct if im wrong.

```
#emissions from fix shape_id
emi_list_7A <- emission_model(tp_model = tp_model_7A_R01,
                               ef_model = "ef_europe_emep"
                               , fleet_data = fleet_data
                               , pollutant = c("CO2","PM10")
                               , reference_year = 2020 )

#analyzing emission estimates
emi_dt <- emis_to_dt(emi_list = emi_list_7A,
                     veh_vars = "veh_type",
                     pol_vars = "pollutant")
emi_dt
```

Thanks for the kind help!

**Reply by Joaobazzo** [reply]
Hi @nadiaqamilla ,
1) Yes, there are multiple `trip_id`s with the same VKT. This happens because all those `trip_id` are associated with the same `route_id`. But you can verify that they start/end at different times - which is common for most of the public transport routes. 
```R
library(gtfs2emis)
library(gtfstools)

# read GTFS
gtfs_file <- system.file("extdata/bra_cur_gtfs.zip", package = "gtfs2emis")
gtfs <- gtfstools::read_gtfs(gtfs_file) 

# keep a single route_id to speed up this example
gtfs_small <- gtfstools::filter_by_route_id(gtfs, "809")

# run transport model
tp_model <- gtfs2emis::transport_model(gtfs_data = gtfs_small,
                                       min_speed = 2,
                                       max_speed = 80,
                                       new_speed = 20,
                                       spatial_resolution = 100,
                                       parallel = FALSE)
# check trip_id's associated with the route '809'
unique(tp_model$trip_id)
length(unique(tp_model$trip_id))

# check when each trip_id starts (view the first obervation of each trip)
setDT(tp_model)[,.SD[1],by = .(trip_id,trip_number)]

# check VKT
setDT(tp_model)[,sum(dist),by= (trip_id)]
```

2) Yes, you can run the model of multiple trips and routes of your GTFS. It will depend on the `transport_model()` function that you previously used. If you apply `transport_model()` for a GTFS that has only one route (see the example in item 1), the associated emissions will be only for the selected trips.  However, if you use the whole GTFS, you can have emissions distributed by shape_id, route_id, trip_id and so on. To verify the emissions you can use some of the post-processing functions of the package, such as `emis_to_dt()`, `emis_summary`, or `emis_grid` (for spatial analysis).  You can see some examples in the [vignette](https://ipeagit.github.io/gtfs2emis/articles/gtfs2emis_intro_vignette.html)

**Comment by rafapereirabr** [comment]
Note that
`gtfs2emis::transport_model()` considers network distances between transit stops (along the transit route geometry). Meanwhile,  `gtfstools::get_trip_length()` uses Euclidean distances between stops. The obvious disadvantage here is that the Euclidean distance approach can substantially underestimate distances between stops and VKT. One advantage here is that the Eucliean distance approach is that it works even for GTFS feeds without the `shapes.txt` data file.
